subprojects/docs/src/docs/userguide/developing-plugins/writing_in_kotlin.adoc

```diff
@@ -0,0 +1,309 @@
+= Writing Gradle plugins in Kotlin
+
+In addition to <<kotlin_dsl.adoc, writing build scripts scripts>>, Gradle also supports writing plugins in https://www.jetbrains.com/opensource/kotlin/[Kotlin].
+
+There are multiple options to write your Kotlin plugins as well as multiple compatibility considerations. This page is non-exhaustive list of dependencies and compiler configurations and their impacts in terms of compatibility and usability.
+
+It also describes the different development plugins to help you set up these different options.
+
+
+[[sec:dependencies]]
+== Different dependencies
+
+Like other JVM projects, Gradle plugins are jar files with dependencies. Unlike other JVM projects, Gradle provides fixed versions of some dependencies at runtime. Those dependencies are `compileOnly` dependencies and need to be compatible with the version of Gradle your plugin is running in.
+
+If you are writing convention plugins, this is usually not an issue. But if you distribute your plugins, you should pay attention to their compatibility.
+
+[[sec:gradle_api]]
+=== gradleApi()
+
+```kotlin
+dependencies {
+    compileOnly(gradleApi())
+}
+```
+
+`gradleApi()` is the primary way of interacting with the https://docs.gradle.org/current/javadoc/index.html[Gradle core].
+
+It is provided by the version of Gradle you're using to build your plugin. It is built the first time you run a build and contains all the symbols in your current version of Gradle. You will usually find it under `~/.gradle/caches/$version/generated-gradle-jars/gradle-api-$version.jar`.
+
+Because `gradleApi()` uses the same version as the one you're currently running, it's easy to use symbols that are not present in previous versions of Gradle. If your plugin is run by an older version that is missing such a symbol, your plugin will throw a `NoSuchMethodError` at runtime.
+
+Note: Because it is provided by Gradle at runtime, `gradleApi()` should be a `compileOnly` dependency.
+
+[[sec:redistributed_gradle_api]]
+=== Redistributed Gradle API
+
+
+```kotlin
+dependencies {
+    compileOnly("dev.gradleplugins:gradle-api:$version")
+}
+```
+
+The https://docs.nokee.dev/manual/gradle-plugin-development.html[Nokee redistributed Gradle API artifacts] are like `gradleApi()` except they allow to specify a version.
+
+Doing this, you're not risking to use removed APIs. On the other hand, you cannot use newer APIs if your plugin is running on a recent version of Gradle.
+
+Note: Because it is provided by Gradle at runtime, `gradle-api` should be a `compileOnly` dependency.
```

Comment:
Some tweaks/tidying up grammar/formatting

````suggestion
=== Redistributed Gradle API

To avoid the implicit link on a Gradle version that comes with using `gradleApi()`, an alternative is to use the https://docs.nokee.dev/manual/gradle-plugin-development.html[Nokee redistributed Gradle API artifacts]. These are like `gradleApi()` except they allow specifying a specific version of the Gradle API.

```kotlin
dependencies {
    compileOnly("dev.gradleplugins:gradle-api:8.0.2")
}
```

TIP: Because it is provided by Gradle at runtime, `gradle-api` must be a `compileOnly` dependency.

This makes the version of the Gradle API being used to build the plugin more explicit, removing the ambiguity avoids the implicit link between `gradleApi()` and the build version of Gradle.

On the other hand, you cannot use newer APIs if your plugin is running on a recent version of Gradle.
````

Reply:
> compileOnly("dev.gradleplugins:gradle-api:8.0.2")

FWIW, I'm always a bit wary of putting hardcoded versions in codeblocks because they do not age well. I guess asciidocs allow to have variables, I'll look into this.

Reply:
I pushed some changes [there](https://github.com/gradle/gradle/pull/24286/commits/41ee911b0719d9f9495800a3b31a85a0d1183ca7). It's a bit funky because it uses the current SNAPSHOT version like `8.2-20230310230000+0000`. But I think in the end it's acceptable because the published docs will use stable.